doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+2. Detecting at built time if a change has occurred to any of our `.proto` definitions. This goes a
+long way to making the migrations "fearless". If someone changes the current proto definitions, the
+old versions, or bumps the `STASH_VERSION` without snapshotting, we can emit helpful build errors.
+3. It uses "off the shelf" crates, the only code we need to write and thus maintain would be a 
+fairly small `build.rs` script.
+4. **Nice to have**, protobuf has a more compact serialization format than JSON, and is more 
+performant. We also already use protobuf in our codebase, so we're not introducing any new 
+concepts.
+
+
+### Part 2: Structuring Migrations
+
+Along with maintaining `const STASH_VERSION` and using a `match` statement as described above we'd
+setup a pattern of defining migrations in their own modules. Specifically we'd introduce the 
+following layout in our repo:
+
+```
+/stash
+    /src
+        /migrations
+            v0_to_v1.rs
+            v1_to_v2.rs
+            v2_to_v3.rs
+            ...
+        ...
+    /proto
+        ...
+```
+
+and our migration code would then look something like this:
+
+```
+const STASH_VERSION: u64 = 3;
+
+async fn migrate(...) -> Result<...> {
+    let version = stash.read_version().await?;
+
+    const NEXT_VERSION = STASH_VERSION + 1;
+    match version {
+        0 => migrations::v0_to_v1::migrate(...).await?;
+        1 => migrations::v1_to_v2::migrate(...).await?;
+        2 => migrations::v2_to_v3::migrate(...).await?;
+        3 => // migrate
+        NEXT_VERSION.. => panic!("Traveled to the future!"),
+    }
+}
+```
+
+There are a few benefits to putting each migration function in their own modules:
+
+1. The only place we would import the generated protobuf code from our snapshots would be in these
+modules. So the only place that would import the `object_v1.proto` would be the `v0_to_v1` and 
+`v1_to_v2` modules. This helps prevent us from depending on the snapshotted versions anywhere
+besides the necessary migration paths.
+2. We promote a pattern of documenting the migration and why it exists in a module doc comment.
+While it's possible to have doc comments on functions, I believe it would be more obvious/easier to
+review module level doc comments.
+
+Each migration function would be provided a `stash::Transaction`, which would allow us to open 
+[`StashCollection`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/transaction.rs#L136)s
+which should provide enough flexibility to facilitate any migration.
+
+> Note: `StashCollection`s do not enforce uniqueness like [`TableTransaction`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/lib.rs#L775-L793)s
+do, some more thought is needed here as to whether or not uniqueness checks are required for
+migrations. They probably are?
+
```

Comment:
Are we able to remove old migrations after they have shipped in a release? IME it is very important for the long term maintainability of a system to be able to retire old migrations, rather than letting them accrete forever.

Reply:
This totally makes sense, and I believe we should be able to remove old migrations. I added a new section on an "initialization" step for the Stash, which should allow us to remove old migrations, and maintain a "minimum stash version" that we support migrating from.